### PR TITLE
[XLA] Make BN variance calc numerically stable

### PR DESCRIPTION
**Pull request by benbarsdell**
- Batchnorm variance was being computed (and tested) using the equation
  Var[X] = E[X^2] - E[X]^2, which is numerically unstable for large
  mean:variance ratios.
- This commit changes the code (and test) to use the numerically stable
  Var[X] = E[(X - E[X])^2], and modifies the test to ensure that
  numerical stability is a requirement.

Note that this is a 2-pass algorithm. I plan to look into implementing a numerically stable 1-pass algorithm at a later time.

### CODE REVIEW

**Comment by blakehechtman** [comment]
This formulation is only more numerically stable if E(X) is far from zero. If E(X) is closer to zero computing 
variance = sum(X^2)-(sum(X)^2) is more numerically stable due to fewer rounding steps. The parallel reduction is very performance critical to all XLA back-ends and I cannot accept this patch without an option to the pass that defaults to enabling the parallel reduction.

**Comment by benbarsdell** [comment]
> This formulation is only more numerically stable if E(X) is far from zero. If E(X) is closer to zero computing variance = sum(X^2)-(sum(X)^2) is more numerically stable due to fewer rounding steps.

While this may technically be true, in practice it appears to me to be inconsequential because the closer E(X) is to zero, the smaller the relative error can be. It's not obvious to me how to formulate a test case where this results in more than O(1e-7) relative error.

In contrast, a large mean:variance ratio can cause the current one-pass algorithm to fail catastrophically. E.g., a ratio of 1e4 can result in a relative error of O(1) or more.

> The parallel reduction is very performance critical to all XLA back-ends and I cannot accept this patch without an option to the pass that defaults to enabling the parallel reduction.

> I expect this change would have a significant negative performance impact on the GPU backend.

These are indeed serious concerns. My thinking was that fixing a potential failure case was more important than the potential perf regression, but if the regression is likely to be significant then it may be best to look into an alternative one-pass algorithm now rather than later.

I'm aware of two single-pass algorithms that don't suffer from catastrophic precision loss:

1. [One-pass shifted](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Computing_shifted_data): Var[X] = E[(X - K)^2] - (E[X] - K)^2, where K is an approximation to the mean (typically the value of the first element of X).
2. [Welford's algorithm](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Parallel_algorithm). This requires a custom reduction operator operating on 3 variables. CUDNN uses this in many cases*.

I'd appreciate your feedback on potentially implementing one of these options instead.

*I should note that CUDNN does not make any guarantees about which algorithms it uses in this context.

**Comment by jlebar** [comment]
> One-pass shifted: Var[X] = E[(X - K)^2] - (E[X] - K)^2, where K is an approximation to the mean (typically the value of the first element of X).

This looks very simple to implement, and I don't expect much of a performance cost, at least on CPU/GPU.

> Welford's algorithm. This requires a custom reduction operator operating on 3 variables. CUDNN uses this in many cases*.

To check my understanding, the notion is that we split the input up into N partitions, with each thread/warp/block processing one partition.  The thread/warp/block computes mean and variance using a stable algorithm, e.g. the two-pass algorithm.  Then we combine as described in that link.

I think XLA can actually represent this using a fusion with one reduction feeding into another.  This is something that the GPU backend should actually be able to codegen today, but the GPU backend will never actually *generate* this fusion.  And who knows about TPUs.  So this whole approach seems much more complicated.

Seems to me we should try option (1) and see if it's good enough?

**Comment by benbarsdell** [comment]
To clarify the Welford algo, here is a simple Numpy implementation:
```
def welford_merge(a, b):
    Na, Ma, M2a = a
    Nb, Mb, M2b = b
    N = Na + Nb
    M = (Ma * Na + Mb * Nb) / N
    M2 = M2a + M2b + np.square(Mb - Ma) * Na * Nb / N
    result = (N, M, M2)
    return result

def welford_mean_variance(x):
    Ndata = np.ones_like(x)
    M2data = np.zeros_like(x)
    N, M, M2 = reduce(welford_merge, zip(Ndata, x, M2data))
    V = M2 / N # or (N - 1) for sample variance
    assert(N == x.size)
    return M, V
```
Could the reduction be implemented using a custom computation for welford_merge?

**Comment by jlebar** [comment]
Ah, I see.

I think one could implement this reduction in XLA:GPU using a compare-and-swap loop.  Like, if you wrote the HLO code for this, it should work.  I suspect it would not be fast, although who knows.

**Comment by blakehechtman** [comment]
For a mean to standard deviation ratio less than 1 the current algorithm is the best. And the error does not exceed 10^-5 until a mean  to standard deviation ratio of of 10^6.

https://dl.acm.org/citation.cfm?id=3221269.3223036

The cases where catastrophic precision loss will happen in the current batchnorm will not produce a variance of zero which will inevitably cause a NaN. Further keeping multiple accumulators largely avoids this cancellation

However using the two pass approach, the shifted approach can produce very small variances which when applied can create excessively large numbers. which can create big problems. I have observed this behavior on multiple occasions when a model other used unfused batchnorm instead of fused batchnorm.

I think the misconception about the goal of batchnorm is really the issue, the goal is not to get the best value of mean and variance, It is to regularize a the activations of a network to enable faster training. Since the operation is linear, the main goal is achieving centering

**Comment by jlebar** [comment]
WRT implementing Welford in XLA, what seems slow is the fact that we have to do a compare-and-swap loop.  (Also, actually, a CAS loop seems tricky because the thing we're CAS'ing on is larger than the atomic size on the chip.)

OTOH I'd expect a deterministic tree reduction (i.e. a reduction that doesn't use atomics) should work fine, if we implemented that.  Like, I don't see us having to implement anything custom just for this.

**Comment by benbarsdell** [comment]
> For a mean to standard deviation ratio less than 1 the current algorithm is the best. And the error does not exceed 10^-5 until a mean to standard deviation ratio of of 10^6.

That result is for doubles not floats. Quote from your reference: "All evaluated implementations use double precision internally".

For floats, it is contradicted strongly by the modified test contained in this PR, here: https://github.com/tensorflow/tensorflow/pull/25115/files#diff-5f84797641cee51dc56ef1fe45b058bbR122
(it's even more clear if the reference calculation is fixed to use either 2-pass or float64).
Using a mean:variance ratio of 1e4 causes the current implementation to produce a variance of **-16** instead of 1.

The reference also states:
"In particular, we demonstrate that the popular textbook equation E[X^2]-E[X]^2 should not be used for computing the variance. Not only has this version a low precision under certain -- not uncommon -- conditions, but there also exist alternatives with much better precision at little additional runtime cost."

> The cases where catastrophic precision loss will happen in the current batchnorm will not produce a variance of zero which will inevitably cause a NaN.

However, as seen in the test, the current implementation can produce a _negative_ variance, which immediately results in NaNs (due to the sqrt).

> However using the two pass approach, the shifted approach can produce very small variances which when applied can create excessively large numbers. which can create big problems. I have observed this behavior on multiple occasions

I'm not sure I follow the cause of this issue, but I'm interested to hear more about it (especially if it crops up in real use-cases).

> the goal is not to get the best value of mean and variance

True, but the variance must still not be catastrophically wrong. (And in general I'd argue that that's really just deep learning being good at overlooking numerical issues).

> I think one could implement this reduction in XLA:GPU using a compare-and-swap loop.

I might have been optimistic in my assessment of implementing Welford in XLA; that sounds trickier than I thought.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @blakehechtman: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by jlebar** [comment]
Unassigning people because currently this PR is not waiting for review.

**Comment by benbarsdell** [comment]
I've changed the algorithm to use the one-pass shifted method, which should retain both numerical stability and performance.

**Comment by blakehechtman** [comment]
Can we make this version an option to this pass that is disabled by default. If so I am fine with the patch otherwise I have measured performance regressions I have no idea what kind of quality impact this will have. I fear many internal users would have to change hyperparameters and golden references.

**Comment by jlebar** [comment]
@benbarsdell thanks for your patience with us on this one.

We have some questions for you about this, but the most important one that we'd like to understand is, what is the motivation for this change?  Are you seeing this hurt in practice, on real models with real data?  If so, which ones, and do you have steps to reproduce?

I ask for a few reasons.

 1. Blake tells me that we have models that will train to *worse* accuracy after this change.

    I admit I'm not as conversant in the mathematical details as he is, but my understanding from him is that in general, weights tend to have a mean close to 0.  Under such circumstances, this algorithm can be worse than what we currently do.

    Indeed, the argument goes, if models *didn't* usually have mean close to 0, then reduced-precision training simply wouldn't work.

 2. This algorithm isn't easy to implement on all platforms.  Indeed even on GPU, one of the things we've been working with you all towards is batchnorm + conv fusion.  In the simple case, we'd compute sum(x) and sum(x^2) in parallel with outputting the results of the conv.  This becomes significantly more complicated if we wanted to implement this algorithm here.

3. I'm leery of a `--better-batchnorm` flag if we can avoid it; no user is going to understand what it means, and it's just going to be another knob that users have to twiddle.

4. I wonder if simply having more parallel accumulators on GPU would solve your problem.  That is, right now we atomicAdd into a single float, but we could instead do a multistep process where we first reduce into multiple floats and then combine them.  In the limit, this becomes a deterministic reduction algorithm, which is something I know folks have been interested in having anyway.

5. Blake suggests that folks can get the behavior of this patch *today* in TF by disabling fused batchnorm.

WDYT?

Sorry again for leaving you hanging.

**Comment by benbarsdell** [comment]
Thank you for the detailed feedback on this, it's very useful.

I made an oversight in the computation of the mean in the latest code, which probably explains the reduced accuracy you're observing. It's easy to fix, but that still leaves the performance concerns you raised.

I'm going to try one more approach to see if the elusive trifecta of accuracy, robustness, and performance can be achieved. I expect to get back to you on it some time next week.

**Comment by benbarsdell** [comment]
Abandoning this due to unforeseen complications. The last thing I tried was the Welford algorithm, but immaturity of the codegen for variadic reductions prevented this from being a viable option at this stage.

The lack of a motivating example means this is not worth further investigation in the short term. Indeed it seems that networks involving BN likely have small mean:variance ratios by construction, and it is difficult to imagine a realistic scenario in which a model will hit this issue during training.
